ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java

```diff
@@ -482,6 +486,60 @@ public void dynamicNodeAdditionsTest() throws Exception {
     helixCluster.upgradeWithNewHardwareLayout(hardwareLayoutPath);
   }
 
+  /**
+   * Tests that if the xid of an InstanceConfig change is greater than the current xid of the cluster manager, then that
+   * change is ignored - both during initialization as well as with post-initialization InstanceConfig changes.
+   */
+  @Test
+  public void xidTest() throws Exception {
+    if (useComposite) {
+      return;
+    }
+    // Close the one initialized in the constructor, as this test needs to test initialization flow as well.
+    clusterManager.close();
+
+    // Initialization path:
+    MockHelixManagerFactory helixManagerFactory = new MockHelixManagerFactory(helixCluster, null);
+    List<InstanceConfig> instanceConfigs = helixCluster.getAllInstanceConfigs();
+    int instanceCount = instanceConfigs.size();
+    InstanceConfig aheadInstanceConfig =
+        instanceConfigs.get(com.github.ambry.utils.TestUtils.RANDOM.nextInt(instanceConfigs.size()));
+    aheadInstanceConfig.getRecord().setSimpleField(XID_STR, Long.toString(CURRENT_XID + 1));
+    HelixClusterManager clusterManager =
+        new HelixClusterManager(clusterMapConfig, hostname, helixManagerFactory, new MetricRegistry());
+    assertEquals(instanceCount - 1, clusterManager.getDataNodeIds().size());
+    for (AmbryDataNode dataNode : clusterManager.getDataNodeIds()) {
+      String instanceName = ClusterMapUtils.getInstanceName(dataNode.getHostname(), dataNode.getPort());
+      assertFalse(instanceName.equals(aheadInstanceConfig.getInstanceName()));
+    }
+
+    // Post-initialization InstanceConfig change:
+    InstanceConfig ignoreInstanceConfig =
+        instanceConfigs.get(com.github.ambry.utils.TestUtils.RANDOM.nextInt(instanceConfigs.size()));
+    String ignoreInstanceName = ignoreInstanceConfig.getInstanceName();
+    ignoreInstanceConfig.getRecord().setSimpleField(XID_STR, Long.toString(CURRENT_XID + 2));
+
+    AmbryReplica ignoreInstanceReplica = null;
+    for (AmbryDataNode dataNode : clusterManager.getDataNodeIds()) {
+      String instanceName = ClusterMapUtils.getInstanceName(dataNode.getHostname(), dataNode.getPort());
+      if (instanceName.equals(ignoreInstanceName)) {
+        ignoreInstanceReplica = clusterManager.getReplicaIds(dataNode).get(0);
+        ignoreInstanceConfig.getRecord()
+            .setListField(STOPPED_REPLICAS_STR,
+                Collections.singletonList(ignoreInstanceReplica.getPartitionId().toPathString()));
+      }
+    }
+    helixCluster.triggerInstanceConfigChangeNotification();
+    // Because the XID was higher, the change reflecting this replica being stopped will not be absorbed.
+    assertFalse(ignoreInstanceReplica.isDown());
```

Comment:
(Optional) I think it would be better to verify that clustermap indeed received the instanceConfig change and `ignoreInstanceReplica` in this instanceConfig was indeed in stopped state. Do you think it's feasible?